### ISSUE TITLE
[no-unused-vars] false positives for code inside `declare` blocks

### ISSUE THREAD

**Issue by kripod** [typescript-eslint/typescript-eslint#2844]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unused-vars": "warn"
  }
}
```

```TS
import "next-auth";

declare module "next-auth" {
  interface User {
    id: string;
    givenName: string;
    familyName: string;
  }
}
```

```TS
import { PrismaClient } from "@prisma/client";

// Namespace declarations are also problematic
declare namespace globalThis {
  let prisma: PrismaClient; // Value can be anything else
}
```

**Expected Result**

No warnings should be emitted.

**Actual Result**

Invalid warnings are emitted inside `declare` blocks, unlike in v4.8.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.9.0` |
| `@typescript-eslint/parser`        | `4.9.0` |
| `TypeScript`                       | `4.1.2` |
| `ESLint`                           | `7.14.0` |
| `node`                             | `14.15.0` |

**Comment by vidaritos** [comment]
Already fixed (I think?), pending release: https://github.com/typescript-eslint/typescript-eslint/pull/2833

**Comment by kripod** [comment]
@vidaritos I saw that PR but don’t think it has anything to do with `declare` block, unfortunately. Thank you for pointing out that fixes are already being made.

**Comment by bradzacher** [comment]
I am unable to reproduce this against master

![image](https://user-images.githubusercontent.com/7462525/101101197-e5a5d780-357c-11eb-91a5-135dca161d3f.png)

Could you please create a repro repo to help me track down the issue?

**Comment by whitetrefoil** [comment]
I've met this issue in my proj:
https://github.com/whitetrefoil/whitetrefoil-checkin-www/tree/temp/typescript-eslint-2844
It's not a minimal repro, but only 2 files are related:

[/src/store/types.ts](https://github.com/whitetrefoil/whitetrefoil-checkin-www/blob/temp/typescript-eslint-2844/src/store/types.ts)
Here declares `RootState`.

[/src/hooks/use-root-selector.ts](https://github.com/whitetrefoil/whitetrefoil-checkin-www/blob/temp/typescript-eslint-2844/src/hooks/use-root-selector.ts)
And it's used here.

**Comment by bradzacher** [comment]
Thanks for that @whitetrefoil 

I've figured it out.
This is a race condition between `naming-convention` and `no-unused-vars`.
If `no-unused-vars` runs first - it works as expected.
But if `naming-convention` runs first - it breaks.

Why?

Because `naming-convention` uses the same utility as `no-unused-vars` does to detect unused variables, and this utility caches the results for each AST (to save duplicate work!)

https://github.com/typescript-eslint/typescript-eslint/blob/aadb39f0ff500ee99ea80e9009ab61283ca9c8cd/packages%2Feslint-plugin%2Fsrc%2Futil%2FcollectUnusedVariables.ts#L39-L43

However - `no-unused-vars` has a few additional selectors it runs in order for it to not report on ambient declarations. `naming-convention` doesn't run these because it doesn't care.

https://github.com/typescript-eslint/typescript-eslint/blob/a8227a6185dd24de4bfc7d766931643871155021/packages%2Feslint-plugin%2Fsrc%2Frules%2Fno-unused-vars.ts#L250-L288

So if `naming-convention` runs first, it'll generate list X, which includes any ambient variables, then `no-unused-vars` will run, and it'll mark the ambient variables as used, but the utility will still generate list X because of caching.

This is an easy fix.

**Comment by kripod** [comment]
Hooray, thank you for figuring this out and fixing it so fast! 🙌
